base/stats.go

```diff
@@ -0,0 +1,747 @@
+	RequestChangesTime          *SgwIntStat `json:"request_changes_time"`
+	RevProcessingTime           *SgwIntStat `json:"rev_processing_time"`
+	RevSendCount                *SgwIntStat `json:"rev_send_count"`
+	RevSendLatency              *SgwIntStat `json:"rev_send_latency"`
+}
+
+type CBLReplicationPushStats struct {
+	AttachmentPushBytes *SgwIntStat `json:"attachment_push_bytes"`
+	AttachmentPushCount *SgwIntStat `json:"attachment_push_count"`
+	DocPushCount        *SgwIntStat `json:"doc_push_count"`
+	ProposeChangeCount  *SgwIntStat `json:"propose_change_count"`
+	ProposeChangeTime   *SgwIntStat `json:"propose_change_time"`
+	SyncFunctionCount   *SgwIntStat `json:"sync_function_count"`
+	SyncFunctionTime    *SgwIntStat `json:"sync_function_time"`
+	WriteProcessingTime *SgwIntStat `json:"write_processing_time"`
+}
+
+type DatabaseStats struct {
+	AbandonedSeqs           *SgwIntStat `json:"abandoned_seqs"`
+	ConflictWriteCount      *SgwIntStat `json:"conflict_write_count"`
+	Crc32MatchCount         *SgwIntStat `json:"crc32c_match_count"`
+	DCPCachingCount         *SgwIntStat `json:"dcp_caching_count"`
+	DCPCachingTime          *SgwIntStat `json:"dcp_caching_time"`
+	DCPReceivedCount        *SgwIntStat `json:"dcp_received_count"`
+	DCPReceivedTime         *SgwIntStat `json:"dcp_received_time"`
+	DocReadsBytesBlip       *SgwIntStat `json:"doc_reads_bytes_blip"`
+	DocWritesBytes          *SgwIntStat `json:"doc_writes_bytes"`
+	DocWritesBytesBlip      *SgwIntStat `json:"doc_writes_bytes_blip"`
+	DocWritesXattrBytes     *SgwIntStat `json:"doc_writes_xattr_bytes"`
+	HighSeqFeed             *SgwIntStat `json:"high_seq_feed"`
+	NumDocReadsBlip         *SgwIntStat `json:"num_doc_reads_blip"`
+	NumDocReadsRest         *SgwIntStat `json:"num_doc_reads_rest"`
+	NumDocWrites            *SgwIntStat `json:"num_doc_writes"`
+	NumReplicationsActive   *SgwIntStat `json:"num_replications_active"`
+	NumReplicationsTotal    *SgwIntStat `json:"num_replications_total"`
+	NumTombstonesCompacted  *SgwIntStat `json:"num_tombstones_compacted"`
+	SequenceAssignedCount   *SgwIntStat `json:"sequence_assigned_count"`
+	SequenceGetCount        *SgwIntStat `json:"sequence_get_count"`
+	SequenceIncrCount       *SgwIntStat `json:"sequence_incr_count"`
+	SequenceReleasedCount   *SgwIntStat `json:"sequence_released_count"`
+	SequenceReservedCount   *SgwIntStat `json:"sequence_reserved_count"`
+	WarnChannelsPerDocCount *SgwIntStat `json:"warn_channels_per_doc_count"`
+	WarnGrantsPerDocCount   *SgwIntStat `json:"warn_grants_per_doc_count"`
+	WarnXattrSizeCount      *SgwIntStat `json:"warn_xattr_size_count"`
+
+	// These can be cleaned up in future versions of SGW, implemented as maps to reduce amount of potential risk
+	// prior to Hydrogen release.
+	CacheFeedMapStats  *ExpVarMapWrapper `json:"cache_feed"`
+	ImportFeedMapStats *ExpVarMapWrapper `json:"import_feed"`
+}
+
+type ExpVarMapWrapper struct {
+	*expvar.Map
+}
```

Comment:
I don't think it can be simplified to that as we have to be able to run a method on the underlying type. Also couldn't come up with a succinct name for the wrapper that would explain its functionality and so I added a comment.

```
// This wrapper ensures that a expvar.Map type can be marshalled into JSON. The expvar.Map has no method to go direct to
// JSON however the map.String() method returns JSON and so we can use this.
```

Reply:
You can define a type based on another type and use it as a receiver - you don't need to define as a struct.  There are several examples in the SG codebase I think - ConflictResolverTrype is a simple example, LogPriorityQueue is another.

Reply:
Yes but we need to call String() on the underlying Map and to do this you have to use a struct.